src/Chain.cpp

```diff
@@ -112,15 +112,55 @@ std::vector<int> Chain::ViterbiPath(int row, int col) const {
 /// column.
 IntVectorVector Chain::ViterbiPaths() const {
   IntVectorVector paths;
-  for (int i = 0; i < viterbi_idx_.rows(); i++) {
-    for (int j = 0; j < viterbi_idx_.cols(); j++) {
+  // Note that we call `viterbi_idx()` (not `viterbi_idx_`) because this ensures
+  // `viterbi_idx_` is computed.
+  for (int i = 0; i < viterbi_idx().rows(); i++) {
+    for (int j = 0; j < viterbi_idx().cols(); j++) {
       paths.push_back(ViterbiPath(i, j));
     }
   }
   return paths;
 };
 
 
+/// @brief If a Chain is clean, mark it as dirty and recursively mark the
+/// downstream Smooshishs as dirty.
+void Chain::MarkAsDirty() {
+  if (!is_dirty_) {
+    is_dirty_ = true;
+    prev_->MarkAsDirty();
+    curr_->MarkAsDirty();
+  }
+};
+
+
+/// @brief If a Chain is dirty, mark it as clean, resize its marginal/viterbi
+/// probability matrices to empty matrices, and recursively mark the downstream
+/// Smooshishs as clean.
```

Comment:
Ah yes, so the way you designed Chain, it checks to see if `marginal_` and/or `viterbi_` exists, and if either of them do, use their scalers and compute the matrices.

When we call `MarkAsClean`, the dirty raw Smooshables have already been fixed with proper `marginal_` matrices, and essentially we want to propagate those changes up through the Chain, and store the newer `marginal_` and `viterbi_` matrices for each upstream Chain.

By making the `marginal_` and `viterbi_` matrices empty for each upstream Chain, whenever we ask for the FIRST marginal/viterbi calculation on any Chain, it'll notice that both matrices are empty, and properly re-compute matrices and scaler counts (now that the raw Smooshables are fixed).

Does this make sense? I chose to leave them empty, purely because I know we'll call `marginal()` and `viterbi()` at some point later, and it'll re-calculate and cache on its own.